[RNMobile] Add new block type badge to block inserter list

dcalhoun opened the pull request:
* **Gutenberg:** Self
* **Gutenberg Mobile:** https://github.com/wordpress-mobile/gutenberg-mobile/pull/3720
* **WPiOS:** https://github.com/wordpress-mobile/WordPress-iOS/pull/16857
* **WPAndroid:** https://github.com/wordpress-mobile/WordPress-Android/pull/15038

## Description
Fixes #30627. Add new block type badge to the native block inserter list. A "new" badge is displayed for block types configured as new. The configuration resides in the JavaScript and is provided to the native host app for storage. Once a given block type is inserted, it is no longer considered new. 

https://github.com/WordPress/gutenberg/issues/33528 is planned to further improve this feature by having publishing multiple posts also remove the "new" badge from block types. That feature is not included in this PR. 

## How has this been tested?
1. Delete previous WP app installations from device. 
2. Install WPiOS/WPAndroid build from sibling PRs. 
3. Log into account and open the Gutenberg Editor. 
4. ℹ️ **Expected:** Block inserter tooltip displayed.<sup>1</sup> 
4. Tap + button to open block inserter. 
5. ℹ️ **Expected:** Observe a "new" badge on the Embed, Search, and Audio blocks. 
6. Inserting one of these blocks. 
7. Reopen the block inserter. 
8. ℹ️ **Expected:** The block inserted no longer has a "new" badge. 
9. Exit the Gutenberg Editor. 
10. Reopen the Gutenberg Editor. 
12. ℹ️ **Expected:** Block inserter tooltip not displayed. 
11. Open the block inserter. 
12. ℹ️ **Expected:** The block inserted still does not have a "new" badge, but the other unused block types do have a "new" badge. 

[1]: ⚠️ The Editor Onboarding features, including the "new" block badge, is only visible for user's within in the "cohort." This consists of 50% of users based upon the user's ID. If the user is not in the cohort, the "new" block badge should not show up. 

## Screenshots 
| iOS | Android |
| --- | --- |
| ![ios-new-block](https://user-images.githubusercontent.com/438664/128043293-446b3868-7356-4c02-98f3-a92611fd3685.PNG) | ![android-new-block](https://user-images.githubusercontent.com/438664/128042132-111011d9-1e6b-42fc-99aa-9d5439e7b006.jpg) |
| ![ios-new-block-dark](https://user-images.githubusercontent.com/438664/128043346-51be74ff-4dde-4492-b4d2-de96be518e68.PNG) | ![android-new-block-dark](https://user-images.githubusercontent.com/438664/128042089-fa336025-85d8-4fd6-80d4-02948bd08f90.jpg) |

## Types of changes
New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

dcalhoun committed:
[wip] Add new block badge

dcalhoun committed:
[wip] Impressions dictate newness and decrement when block type inserted

The new indicator is now toggled based upon the impression count for a
block type. The impression count is decremented any time the block type
is inserted.

dcalhoun committed:
De-structure arguments to simplify passing arguments to parent callback

Capturing all the arguments in a single assignment simplifies passing
them to the parent callback.

dcalhoun committed:
Request and set block type impressions from native iOS

dcalhoun committed:
Rename native bridge methods for clarity

Avoid confusion from similarly named methods between React Hooks and
native bridge methods.

dcalhoun committed:
Request and set block type impressions from native Android

dcalhoun committed:
Remove unused file

This was erroneously added during exploration.

dcalhoun committed:
Add getSettings mock for failing unit test

We must mock this newly referenced property in the BlockTypesTab store
selector.

dcalhoun committed:
Add mocks for native bridge methods to fix failing test

Avoid unexpected test failures from missing methods.

SiobhyB commented on packages/components/src/mobile/inserter-button/index.native.js:
```diff
@@ -80,6 +87,9 @@ class MenuItem extends Component {
 							isClipboardBlock && clipboardBlockStyles,
 						] }
 					>
+						{ item.isNew && (
```
👋 @dcalhoun, I was just curious whether there might be an opportunity to use [the badge component that was recently created](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/mobile/badge/README.md) for featured images here? I checked and see that wrapping the `<Badge>` component around `<View style={ modalIconStyle }>` (as well as passing a smaller `margin` and `font-size`) yields similar results for the badge: 

![Screenshot_1625153118](https://user-images.githubusercontent.com/2998162/124150082-21f4d400-da89-11eb-9f68-01c7c5fe98c0.png)

This is definitely a question (rather than an informed opinion) on my part, I'm curious to hear whether there'd be disadvantages to using that component vs. the current approach. :)

dcalhoun replied:
Thank you for the suggestion, @SiobhyB! I agree that reusing a shared badge could make a lot of sense for consistency and avoid duplicative code. 

I had actually _briefly_ tried your `Badge` component while I was building, but ran into the positioning and sizing discrepancies you implied. Passing custom `margin` and `font-size` could address that, but I wonder if we are attempting to prematurely force an abstraction (i.e. is there a single badge style that we'd like to apply everywhere or would adding diverging styling options make the abstraction unnecessarily complex)? Additionally, I noticed that the `Badge` component doesn't have any dark-mode specific styles, and this block type list context may need those. So, it may require some coordination to see if the badge styles do align well between this block types list and the featured image. 

All that said, I'll definitely take another look to see if sharing the code makes sense. 

@kyleaparker WDYT about aligning the styles between the featured image and block type list badges? It would appear the padding, font size, background, and border radius may be a little bit different, as the block type list is more space constrained and requires a smaller badge. Do we have a desire to standardize this badge or should we treat the block type list as a separate pattern? Additionally, do we need dark-mode styles for the badge? 

| Featured Image | Block Type List |
| --- | --- |
| <img width="388" alt="featured-image-badge" src="https://user-images.githubusercontent.com/438664/124153093-b2b3bb80-da59-11eb-8bd2-d5619c5007d2.png"> | <img width="599" alt="block-type-list-baddge" src="https://user-images.githubusercontent.com/438664/124153132-b8a99c80-da59-11eb-8cf8-748594a1e83c.png"> |

SiobhyB replied:
Thanks, David! Looking forward to following the discussion here and would be happy to help in any way if there's a decision to align the badge styles. I also wanted to note that the padding in the "featured" badge has been reduced since that purple screenshot, following design feedback in https://github.com/WordPress/gutenberg/pull/32865, here's an updated screenshot for comparison:

![122811271-2c181500-d2c8-11eb-9f3f-1b9604877b29](https://user-images.githubusercontent.com/2998162/124255583-3c2ec080-db22-11eb-9759-f5245c0d44e9.png)

kyleaparker replied:
@dcalhoun let's align on the styles - I think it's better they match :)

dcalhoun replied:
> I also wanted to note that the padding in the "featured" badge has been reduced since that purple screenshot, following design feedback in #32865 [...]

@SiobhyB ah, thank you for sharing that. I was unaware of that. It likely changed since the first time I explored reusing the `Badge` component. 

> @dcalhoun let's align on the styles - I think it's better they match :)

@kyleaparker sounds good. I suppose I'll use my discretion in modifying the styles to work in both contexts and dark mode; and let you know if I run into any specific issues.

dcalhoun committed:
Abstract block type impression tracking

Avoid duplicative code between the block type tab and the block type
search results. This abstraction is one approach, likely worth iterating
on further.

dcalhoun committed:
Set default impression count for core/paragraph

Impression count set for debugging purposes.

dcalhoun committed:
Limit impression decrement function scope

We only need to use this function when decreasing the impression count
when a block has been inserted. So, we can make the name clearer by
simplifying its logic.

dcalhoun committed:
Relocate impressions state to block editor store

Relocating the state prevents unnecessary trips across the native bridge
requesting the impression counts. Now, one request is made when the
editor session begins. The impression counts are stored in the block
editor store for easy access from within the block inserter.

dcalhoun committed:
Abstract trackBlockTypeSelection argument de-structuring

This avoids a small amount of duplicative code. The method signature is
still clear from my perspective.

dcalhoun committed:
Leverage shared Badge component

Improve UI consistency and avoid duplicative code through reusing an
existing Badge component. Additional props were required to make the
Badge adaptable to the space-constrained block inserter. Lastly, the
Badge mark up was simplified and unused code was removed.

dcalhoun committed:
Set new block types from JavaScript

Relocated defining new block types to be co-located with the
registration of block types.

dcalhoun committed:
Fix erroneous new badge on blocks

Previously, the `isNew` prop would be set to `undefined` when the editor
onboarding capability was disabled. This caused the `Badge` component to
fallback on its default configuration to be shown. By explicitly
checking for a `true` value, the badge is appropriately hidden when
editor onboarding is disabled.

dcalhoun committed:
Refactor Android bridge methods to scope types

Avoid "leaking" React Native types deeper into the WPAndroid source by
performing the necessary conversions within the glue code.

dcalhoun committed:
Improve NEW_BLOCK_TYPE documentation

dcalhoun committed:
Update which block types are considered "new"

These three blocks types are the latest to be supported.

dcalhoun committed:
Fix typo in parameter name

dcalhoun committed:
Refactor Android impression count Integer to Double

This better aligns with what React Native expects and what Gson produces
when stringifying JSON. Previously, there would be mismatch type errors
between Double and Integer.

dcalhoun commented on packages/block-library/src/index.native.js:
```diff
@@ -277,3 +277,14 @@ export const registerCoreBlocks = () => {
+ *
+ * @constant {string[]}
+ */
+export const NEW_BLOCK_TYPES = [ embed.name, search.name, audio.name ];
```
There are the last three blocks for which we added mobile support. Open to thoughts on which blocks we should mark as "new."

dcalhoun commented on packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/GutenbergBridgeJS2Parent.java:
```diff
@@ -164,4 +167,8 @@ void gutenbergDidRequestUnsupportedBlockFallback(ReplaceUnsupportedBlockCallback
 
+    void requestBlockTypeImpressions(ReadableArray newBlockTypes, BlockTypeImpressionsCallback blockTypeImpressionsCallback);
+
+    void setBlockTypeImpressionCount(String name, Double count);
```
`Double` is used to satisfy both [React Native's expectations](https://reactnative.dev/docs/native-modules-android#argument-types) and the fact that Gson, which is used in WPAndroid to stringify the impressions object, [deserialization appears to automatically use `Double` for all numbers](https://stackoverflow.com/questions/24926786/make-gson-deserialize-numbers-to-integers-or-doubles).

mchowning replied:
Might be worth adding this as a comment? I don't feel strongly though.

mchowning replied:
> and the fact that Gson, which is used in WPAndroid to stringify the impressions object, deserialization appears to automatically use Double for all numbers

Just as an FYI, if GSON was the only issue, I think we could change that to return Integers (for example) with something like this:

```
Map<String, Integer> impressions = GSON.fromJson(jsonString, new TypeToken<Map<String, Integer>>(){}.getType());
```

That obviously doesn't address the "React Native's expectations" aspect of this. I just wanted to make it clear that GSON's default shouldn't be a blocker here.

dcalhoun replied:
Ah, interesting. Thank you for sharing that.

dcalhoun committed:
Split tracking of editor onboarding capability and initial session

Previously, the `editorOnboarding` capability included both (1) whether
the user was within the cohort and (2) if the user had previously
launched the Gutenberg editor. The latter is utilized to only show the
initial tooltip once. However, the combination resulted in the latter
preventing the new block badges from displaying past the initial
session.

Splitting these two attributes allows for features like the new blocks
badge to remain active past the user's initial editor session.

dcalhoun committed:
Remove erroneous line break

kyleaparker commented:
Nice work, this is looking good! I have a few thoughts based on the screenshots: 
- On Android, the badge is getting quite close to some of the icons. Do we have any flexibility with the size?
- On dark mode can we try using the same colour as the badge icons? This would match how it looks on light mode.  
- Is it possible to match the badge border radius with the block border radius?

jhnstn commented on packages/block-editor/src/components/inserter/hooks/use-block-type-impressions.native.js:
```diff
@@ -0,0 +1,51 @@
+	const { updateSettings } = useDispatch( blockEditorStore );
+
+	const items = enableEditorOnboarding
+		? blockTypes.map( ( b ) => ( {
```
Minor non blocking nitpick: would you consider a longer name for the element in callback ? Perhaps the singular of the collection: `blockType`

dcalhoun replied:
@jhnstn definitely. I intended to rename this to avoid the cryptic name, but forgot. Thank you for calling this out.

mchowning commented on packages/editor/src/components/provider/index.native.js:
```diff
@@ -161,6 +163,11 @@ class NativeEditorProvider extends Component {
+		// Request current block impressions from native app
+		requestBlockTypeImpressions( ( impressions ) => {
+			updateSettings( { impressions } );
+		}, NEW_BLOCK_TYPES );
```
It feels like we should be specifying the starting value of `3` for the "impression count" for new blocks somewhere on the js-side, like here (or maybe it should be a part of `NEW_BLOCK_TYPES`), instead of doing it separately on [Android](https://github.com/wordpress-mobile/WordPress-Android/blob/0ef7ba7cfccd38dd99b10040fa78b61f856fffa0/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java#L3450) and [iOS](https://github.com/wordpress-mobile/WordPress-iOS/blob/d28d85ae0d96cd2423d7709addd357046c8b20fe/WordPress/Classes/ViewRelated/Gutenberg/GutenbergViewController.swift#L946). Is there any reason we would ever want Android and iOS to have different values?

mchowning replied:
Could we take that further and actually merge the new block types into the pre-existing impressions inside the callback? That way we would do the merging logic one time instead of separately on Android and iOS. In essence, I think that the main apps should only be persisting the object passed to it and retrieving that same object when requested. I think we could move all the other logic to the JS-side and share it between Android and iOS.

mchowning replied:
On a semi-related note, this has got me wondering again whether we would benefit from giving the main apps a more generic interface for Gutenberg to persist and retrieve key-value pairs along these lines:

```
void getPersistedString(String key, Consumer<String> persistedStringCallback);
void persistString(String key, String value, PersistanceScope persistanceScope);
enum PersistenceScope { User, Global, Site }
```

That's not something we need to tackle in this PR though.

dcalhoun replied:
@mchowning great thoughts. Thank you for the review. 

> It feels like we should be specifying the starting value of `3` for the "impression count" for new blocks somewhere on the js-side, like here (or maybe it should be a part of `NEW_BLOCK_TYPES`), instead of doing it separately on [Android](https://github.com/wordpress-mobile/WordPress-Android/blob/0ef7ba7cfccd38dd99b10040fa78b61f856fffa0/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java#L3450) and [iOS](https://github.com/wordpress-mobile/WordPress-iOS/blob/d28d85ae0d96cd2423d7709addd357046c8b20fe/WordPress/Classes/ViewRelated/Gutenberg/GutenbergViewController.swift#L946). Is there any reason we would ever want Android and iOS to have different values?

No, I do not believe there is an explicit need for the platforms to diverge at this time. I believe my initial implementation likely erred on the side keeping the JavaScript API simple. Providing "control" to JavaScript and avoiding duplicative native code likely makes more sense. I will refactor the work to accommodate this. 

> Could we take that further and actually merge the new block types into the pre-existing impressions inside the callback? That way we would do the merging logic one time instead of separately on Android and iOS. In essence, I think that the main apps should only be persisting the object passed to it and retrieving that same object when requested. I think we could move all the other logic to the JS-side and share it between Android and iOS.

If I follow your proposal correctly, I originally opted against that approach to avoid a second trip over the native bridge, once to retrieve the persisted values and once to set the persisted values based upon `NEW_BLOCK_TYPES` configuration. 

Currently, the [impressions are seeded](https://github.com/WordPress/gutenberg/blob/30307f3e074b39a8423cfe6120fd43c945dc1324/packages/editor/src/components/provider/index.native.js#L167-L170) and only [set individually by name](https://github.com/WordPress/gutenberg/blob/30307f3e074b39a8423cfe6120fd43c945dc1324/packages/block-editor/src/components/inserter/hooks/use-block-type-impressions.native.js#L43-L44) when a given block is inserted. This could be refactored so that the persisted value is set by passing the a new object instead, but we would also need to immediately set the impressions within the request callback. **Do you think deduplicating code is worth the second bridge trip?** 

```js
requestBlockTypeImpressions((storedImpressions) => {
  // `NEW_BLOCK_TYPES` would be updated to be an object with default a count,
  // e.g. `{ 'core/paragraph': 3 }`
  const impressions = { ...storedImpressions, ...NEW_BLOCK_TYPES };
  // Persist to native store
  setBlockTypeImpressions(impressions);
  // Persist to JavaScript store
  updateSettings(impressions);
});

```

> On a semi-related note, this has got me wondering again whether we would benefit from giving the main apps a more generic interface for Gutenberg to persist and retrieve key-value pairs along these lines:
> 
> ```
> void getPersistedString(String key, Consumer<String> persistedStringCallback);
> void persistString(String key, String value, PersistanceScope persistanceScope);
> enum PersistenceScope { User, Global, Site }
> ```
> 
> That's not something we need to tackle in this PR though.

Yes, some form of key-value store does seem to be a very common requirement for many features. Building an abstraction to support this frequent requirement does make sense IMO. Surely at this point we have enough use cases to identify a common API for the temporary persistence layer JavaScript requires. Given the elongated timeline of this project, I would prefer to attempt that abstraction in a different PR/project.

mchowning replied:
> some form of key-value store does seem to be a very common requirement for many features. Building an abstraction to support this frequent requirement does make sense IMO. Surely at this point we have enough use cases to identify a common API for the temporary persistence layer JavaScript requires. Given the elongated timeline of this project, I would prefer to attempt that abstraction in a different PR/project.

👍  This does not need to be done as a part of this project.

mchowning replied:
> > If I follow your proposal correctly, I originally opted against that approach to avoid a second trip over the native bridge, once to retrieve the persisted values and once to set the persisted values based upon NEW_BLOCK_TYPES configuration.

Could we make the extra trip a very rare occurrence? For example, I'm wondering if we could do something like:

1. Fetch impressions from native
2. On the JS side, check if the impressions include the blocks from `NEW_BLOCK_TYPES`. Only if the `NEW_BLOCK_TYPES` are missing do we "update" the impressions with the `NEW_BLOCK_TYPES` impressions (so this "second" call would only happen one time for each change to the `NEW_BLOCK_TYPES`). 

> Currently, the impressions are seeded and only set individually by name when a given block is inserted. This could be refactored so that the persisted value is set by passing a new object instead, but we would also need to immediately set the impressions within the request callback.

I think this could make sense, particularly with the flow I suggested above since it would avoid having to "set" the impressions for _each_ new block type. In fact, it might even enable us to avoid having any extra trips across the bridge--we could let the `NEW_BLOCK_TYPES` get "added" to the impression count on the native side as a part of the update that would occur whenever the number of impressions for one of the blocks gets updated (since we are now sending the entire "object," including the added-in-the-callback `NEW_BLOCK_TYPES` across with that update). 

The one concern with sending the full object that comes to mind for me is whether that object might grow over time to be so large that its size could impact performance (relative to just sending an update with a single key/value). That feels like a very hypothetical concern. I would love for us to have so many blocks that this object became too big, but I don't see it happening anytime in the remotely foreseeable future. Also, if we really wanted to, I guess we could avoid that size issue by just passing a map of updates instead of overwriting the entire object.
 
> Do you think deduplicating code is worth the second bridge trip?

I do, but I also think it is a point that reasonable people could disagree on, so don't hesitate to let me know if you do. 🙂

dcalhoun replied:
> Could we make the extra trip a very rare occurrence? 

@mchowning yes, I agree that a check within the callback for newly added block types could allow us to conditionally perform the second trip as needed. 

> In fact, [persisting the value with a new object] might even enable us to avoid having any extra trips across the bridge--we could let the `NEW_BLOCK_TYPES` get "added" to the impression count on the native side as a part of the update that would occur whenever the number of impressions for one of the blocks gets updated (since we are now sending the entire "object," including the added-in-the-callback `NEW_BLOCK_TYPES` across with that update).

Clever. I like it. I do have concern that it's possibly _too clever_, where it might cause confusion for a future contributor modifying this code, but I like it enough to explore. Hopefully I can identify a pattern/naming that makes it very clear what is occurring. 

> The one concern with sending the full object that comes to mind for me is whether that object might grow over time to be so large that its size could impact performance (relative to just sending an update with a single key/value). 

That makes sense. I agree this becoming a significant concern is likely a ways off. 

> > Do you think deduplicating code is worth the second bridge trip?
> 
> I do, but I also think it is a point that reasonable people could disagree on, so don't hesitate to let me know if you do. 🙂

I do not have a staunch opinion on the matter. I will explore postponing native persistence until a block is inserted. If that proves confusing, I will explore a conditional second bridge trip.

mchowning replied:
>  I do have concern that it's possibly too clever, where it might cause confusion for a future contributor modifying this code

👍  Definitely a fair concern. It might be worth just having the occasional extra trip across the bridge when `NEW_BLOCK_TYPES` changes if that allows us to make the code easier to understand.

mchowning replied:
I was thinking about this last night, and one additional reason to favor moving logic to js from the native side (apart from avoiding duplicated logic) is that it would make it easier for other apps to integrate gb-mobile. There will be less complexity they will need to add if we keep as much complexity as possible "inside" gb-mobile.

dcalhoun commented:
Thank you for the review, @kyleaparker. 

> * On Android, the badge is getting quite close to some of the icons. Do we have any flexibility with the size?

We can definitely modify the size. When attempting to infer the text sizing and spacing values from the design comps, I began to grow concerned about legibility. Currently the "small" variant of the badge used in the block inserter relies upon 12px type. I can lower that further and potentially the padding in an attempt to avoid collision with the icons. Let me know if there specific values you'd like to see. 🙇🏻 

> * On dark mode can we try using the same colour as the badge icons? This would match how it looks on light mode.

I interpret "try using the same colour as the badge icons" to mean have the "new badge" background match the block type button icon color and the "new badge" text color match the block type button background color. I will give this a try. Please let me know if I misunderstood. 

> * Is it possible to match the badge border radius with the block border radius?

Yes, we can definitely modify the badge border radius. I did notice that your original design appeared to mimic the parent block border radius. However, now that we are sharing a `Badge` component between this feature that Featured Image, modifying it will impact both places. 

Alternatively we could modify _only the "new" badge_ border radius. That would be simple enough to accomplish now, but I do have concerns about applying a one-off modification like that to a component we are attempting to standardize, i.e. that was [my initial concern](https://github.com/WordPress/gutenberg/pull/33119#discussion_r662427421) about prematurely abstracting/sharing the `Badge` component. 

WDYT? How would you like to approach the "new" badge border radius? 

---

**Update:** I implemented the above changes via (1) reducing the padding, (2) adding a dark mode variant, and (3) increasing the border radius from the `Badge` universally. @kyleaparker will you please review the below images and let me know if this aligns better with what you envision? 


<details><summary>Screenshots of Style Changes</summary>

| iOS | Android |
| --- | --- |
| <img alt="ios-1" src="https://user-images.githubusercontent.com/438664/127186173-8dd9578b-5ecf-4f0f-a667-03b4bf7e268e.png"> | ![android-1](https://user-images.githubusercontent.com/438664/127186212-5dbd7a6f-ca13-4dfc-8e5c-626fefe32811.jpg) |
| <img alt="ios-2" src="https://user-images.githubusercontent.com/438664/127186238-a6cae04b-04c7-428e-b921-138cec9b464f.png"> | ![android-2](https://user-images.githubusercontent.com/438664/127186262-30316d9d-ee17-409b-86b2-de2b15bd50ec.jpg) |
| <img alt="ios-3" src="https://user-images.githubusercontent.com/438664/127186284-90e4cf93-7144-49ab-b686-a9345bb9af6c.png"> | ![android-3](https://user-images.githubusercontent.com/438664/127186318-4db988d2-e1fd-4cb1-be7c-402770ec9e23.jpg) |

</details>

dcalhoun committed:
Rename parameter to improve clarity

Single letter parameter names may cause confusion.

dcalhoun committed:
Reduce padding of small badge variant

Reduce badge size to avoid collisions with block type icons.

dcalhoun committed:
Add dark mode badge variant

Improve UI style consistency via dark mode support for badge.

dcalhoun committed:
Increase badge border radius to match block inserter button

Improve styling consistency between block inserter button and badge.

guarani commented on packages/react-native-bridge/index.js:
```diff
@@ -396,4 +396,32 @@ export function requestPreview() {
+ * Request the host app set updated impression count for a given block type identified by name.
+ *
+ * @param {string} name  The block type name to update.
+ * @param {number} count The number of impressions for the block type.
```
I'm not sure where to leave this comment, but I wanted to raise how the meaning given to the word "impression" in this project doesn't feel intuitive to me. Apologies for not raising this earlier when it was first defined in an internal post.
My understand after reading the code and [comments](https://github.com/WordPress/gutenberg/pull/33119/files?file-filters%5B%5D=.java&file-filters%5B%5D=.js&file-filters%5B%5D=.m&file-filters%5B%5D=.swift#diff-3286cc974baacc1ef30a6c7d6c18ae3ea400c86e1223c88029c1ea15051d65ecR282-R286) above is that each block is given a default "impression count" that could be zero or any positive integer. If the count is zero, the block is not considered "new" and no badge is applied. If the count is greater than zero however, the block _is_ still considered "new" and a badge is shown.

The closest definition of the word "impression" that seems to apply to the context here is the marketing term which I understand as "an impression occurs when something has been eyeballed by the user". Given this point of view, I expected the impression count to _increase_ each time a block presented to the user (be it in the block inserter or the editor itself). It does the opposite though, which leads to the confusion I had.

What I see here looks more like a "novelty factor", although that's doesn't describe it well either. 
Have we considered flipping the way it's used by giving each of the [new blocks](https://github.com/WordPress/gutenberg/pull/33119/files?file-filters%5B%5D=.java&file-filters%5B%5D=.js&file-filters%5B%5D=.m&file-filters%5B%5D=.swift#diff-3286cc974baacc1ef30a6c7d6c18ae3ea400c86e1223c88029c1ea15051d65ecR290) a "usage score"? For example,
- a view in the block inserter (i.e. the block is displayed to the user): plus 2 points
- an insert (i.e. the block is inserted by the user) via the block inserter or slash inserter: plus 5 points
- a view in a post (i.e. the user opened a post in the editor and it contained the block): plus 1 points 

Only new blocks would have a score (blocks not in `NEW_BLOCK_TYPES` wouldn't have a score, which is how the current implementation works I believe).

guarani replied:
Of course, we'd define a threshold score and when a block's score surpasses this, the badge would no longer be shown.

dcalhoun replied:
> I'm not sure where to leave this comment, but I wanted to raise how the meaning given to the word "impression" in this project doesn't feel intuitive to me. 

> The closest definition of the word "impression" that seems to apply to the context here is the marketing term which I understand as "an impression occurs when something has been eyeballed by the user". 

@Guarani thank you for raising this. I agree the use of the term "impression" is leaning on a loose definition. In reality, it currently represents remaining "actions" before a block is stale. E.g. "inserting" immediately marks a block type stale, #33528 plans to incrementally mark a block type stale when a post is published. As you mentioned, we could also incrementally remove newness when a block type is viewed within the inserter or editor. 

I am definitely open to renaming it to improve clarity. I have found identifying a clear name to be challenging given the various actions that modify newness. Would `noveltyScore` or `newnessRating` better communicate the intent? 

> [...] Given this point of view, I expected the impression count to _increase_ each time a block presented to the user (be it in the block inserter or the editor itself). It does the opposite though, which leads to the confusion I had.
>
> What I see here looks more like a "novelty factor", although that's doesn't describe it well either.
> Have we considered flipping the way it's used by giving each of the [new blocks](https://github.com/WordPress/gutenberg/pull/33119/files?file-filters%5B%5D=.java&file-filters%5B%5D=.js&file-filters%5B%5D=.m&file-filters%5B%5D=.swift#diff-3286cc974baacc1ef30a6c7d6c18ae3ea400c86e1223c88029c1ea15051d65ecR290) a "usage score"? For example,
> 
> * a view in the block inserter (i.e. the block is displayed to the user): plus 2 points
> * an insert (i.e. the block is inserted by the user) via the block inserter or slash inserter: plus 5 points
> * a view in a post (i.e. the user opened a post in the editor and it contained the block): plus 1 points

In regards to the current inverted nature of the count, that was done for simplicity. Specifically, I think it is expected/easy to communicate the counter will stop decrementing when it reaches zero. If we count up, we'd need to define an arbitrary maximum for the count. If counting down is truly confusing, we could flip it and define a maximum. 

One alternative could be renaming to communicate that the value counts downward. That could be one of the aforementioned names to avoid the "impressions" term (e.g. `noveltyScore`) or a prefix to communicate the value will be reduced (e.g. `remainingImpressions`, `availableImpressions`, `unusedImpressions`). WDYT? 

> Only new blocks would have a score (blocks not in `NEW_BLOCK_TYPES` wouldn't have a score, which is how the current implementation works I believe).

From my perspective this relates to the simplicity of the decrementing down to zero. Checking if a given block type count is less than zero is more straightforward than checking if a block type has a score (1) set _and_  (2) less than `<arbitary-maximum>`. That is likely a personal opinion, and not one that I hold onto strongly. WDYT? 

I'd also be curious to see @jhnstn's thoughts, as he was the original source of the impression count idea.

mchowning replied:
>  I expected the impression count to increase each time a block presented to the user (be it in the block inserter or the editor itself). It does the opposite though, which leads to the confusion I had.

Thanks for mentioning this @guarani . I had the same confusion, meant to bring it up, and promptly forgot to do that. 😞 

>  I have found identifying a clear name to be challenging given the various actions that modify newness. 

<nod><nod> I suspect that my inability to think of a good name is one of the reasons I conveniently "forgot" to bring this up. 😆 

> In regards to the current inverted nature of the count, that was done for simplicity. Specifically, I think it is expected/easy to communicate the counter will stop decrementing when it reaches zero. If we count up, we'd need to define an arbitrary maximum for the count. If counting down is truly confusing, we could flip it and define a maximum.

I also think that decrementing the count makes sense. 

> One alternative could be renaming to communicate that the value counts downward. That could be one of the aforementioned names to avoid the "impressions" term (e.g. noveltyScore) or a prefix to communicate the value will be reduced (e.g. remainingImpressions, availableImpressions, unusedImpressions). WDYT?

The prefix names all seem like nice improvements. We could even expand them to be even more descriptive with something like `remainingNewBlockImpressions`. I do not have any strong feelings on the specific name though.

jhnstn replied:
I would stick with impressions as the term but I can see where there might be some confusion. Specifically I see how decrementing an `impression` value seems odd if encountering a snapshot of the system i.e. I see the state of an app install and a block has an impression count of `N`. Without knowing the semantics of "impression" in this context I could conclude that the block has made `N` impressions. I still wouldn't know what the impression threshold is ( or if one exists at all) but assuming impression count going up is plausible.  

Personally I don't see it this way. I see it as the block _having_ a set amount of impressions to give. Therefore I would assume that count would be going down. I am biased since I already envisioned this system based on prior experience where a "newness" value goes down.

That said the term is too subjective to know by looking purely at the data store to know which way the system moves that value. In either case it would behoove us to reify that in the name such as : `impressionsMade` or `impressionsRemaining`. Right away I don't have to spell out which term should be used in an incrementing or decrementing system.  I have no preference on the actual name, prefix or suffix, or what not. 

Naming and caching, all the tough stuff.

I would also stick with a decrementing system since it's simpler and less prone to accidentally making something old looking new again.

guarani replied:
Thanks everyone for sharing their thoughts. I agree that there's no clear way to improve this. I don't feel strongly that this _needs_ improving either, so I'm happy to leave it as-is.
I think my concern about it being unintuitive still stands in my mind, but it's hard to name things in a way that lends itself to the correct interpretation to everyone.
While renaming to `remainingImpressions` is probably an option, but not sure it's worth it tbh. So I'm happy to leave this as-is for now.

guarani replied:
I'm happy to mark this thread resolved, but want to give @mchowning a chance to reply first since he also shared thoughts here.

mchowning replied:
I would rename the variable to something like `remainingImpressions`, but I do not view it as a blocker. I'm fine resolving this thread if you'd rather leave it as-is.

dcalhoun replied:
If I am honest, I understand the confusion with the term “impression,” but I do not believe we have an alternative name that meaningfully reduces confusion, at least not enough to rename all the values across three repos at this time. 

I spoke with @mchowning via chat regarding this just now. Based on that chat and the most recent comments from other team members in this thread, I am marking this thread resolved and will leave the implementation as-is. Please feel free to reopen this discussion if any of you feel this is a mistake. 

Thank you. 🙇🏻

guarani commented on packages/components/src/mobile/inserter-button/index.native.js:
```diff
@@ -80,13 +81,20 @@ class MenuItem extends Component {
-							/>
-						</View>
+						<Badge
+							label={ __( 'New' ) }
```
I'd read an [earlier comment](https://github.com/WordPress/gutenberg/pull/33119#issuecomment-887003936) about concerns that the badge would overlap the icon. Is this more of a concern when dealing with other languages where the translation for "New" could be a longer string?

I tested out Spanish by explicitly changing this line to "Nuevo" and the badge overlaps the icon significantly. A few alternatives came to mind which I tested out and added screenshots below. 

| Alternative | Screenshot |
| - | - |
| Current (in Spanish) | <img width="300" alt="Screen Shot 2021-07-27 at 19 38 29" src="https://user-images.githubusercontent.com/1898325/127241331-b9c72b14-74f2-4981-819b-0ad2e3b81f11.png"> |
| Emoji badge | <img width="300" alt="Screen Shot 2021-07-27 at 19 46 08" src="https://user-images.githubusercontent.com/1898325/127241946-d73e08e4-b7ef-432d-aa09-ca29b5f74ed0.png"> |
| Emoji in block title | <img width="300" alt="Screen Shot 2021-07-27 at 19 49 27" src="https://user-images.githubusercontent.com/1898325/127241961-44af6fc3-f195-4d78-bf1c-f477ac81680f.png"> |
| Emoji and text in block title (in Spanish) | <img width="300" alt="Screen Shot 2021-07-27 at 19 59 41" src="https://user-images.githubusercontent.com/1898325/127242558-ba4423c3-e4a1-4dc6-b11c-14cbbb91ff6b.png">

dcalhoun replied:
@guarani this is a great point, and one that I had not considered. Another alternative — similar to the emoji — could be a simple "unread" indicator, much like iOS' Messages app.

![iOS Messages Unread Indicator](https://user-images.githubusercontent.com/438664/127244484-64bf7b76-3a7c-4ea2-89ea-c9a695acdbe3.jpeg)

@kyleaparker curious if you have thoughts on this issue regarding i18n and UI collisions.

mchowning replied:
One related point is that the current new flag is not very visible in dark mode. I realize you probably just haven't gotten to the point of addressing this yet, but I wanted to raise it since I did notice it.

<img src="https://user-images.githubusercontent.com/4656348/127327157-e18de434-df70-4201-b7b3-d401f7314e8f.png" width="400px"/>

dcalhoun replied:
> One related point is that the current new flag is not very visible in dark mode. 

Thanks @mchowning. I updated https://github.com/WordPress/gutenberg/pull/33119#issuecomment-887003936 to include screenshots of the updated dark mode styles. Awaiting @kyleaparker's feedback.

mchowning replied:
Ah yes, sorry I missed that! Those do look much better! 😄

kyleaparker replied:
Good point about other languages causing more problems. My preference would be to go with the star emoji badge on the top left corner. It's a little ambiguous but our main intention is to let people know about new badges and the star will draw people's attention. 

We have some upcoming design updates to the inserter that should give us more space to work with, but I think this is a good compromise for now.

dcalhoun committed:
Relocate default count and merging of impression objects to JavaScript

This refactor de-duplicates and simplifies native host integration by
placing more logic in JavaScript.

guarani reviewed (commented):
👋 Nice work on this so far, @dcalhoun! I left a couple of comments and will be happy to do another pass whenever you like ton finish reviewing.

dcalhoun committed:
Add note regarding required Double type for Android

Double utilized to satisfy both React Native's expectations
(https://bit.ly/3f3tsT4) and Gson's deserialization which uses Double
for all numbers (https://bit.ly/2VfpvUv).

dcalhoun committed:
Replace new indicator text badge with image

To avoid issues with i18n and UI collisions, we opted to use an emoji
image as the new indicator.

dcalhoun committed:
Revert all changes to Badge component

This work no longer depends upon the Badge component, so the
modifications to it are no longer relevant to this branch.

dcalhoun committed:
Change translator note to improve consistency

Other translator notes utilize the term "name" when referring to blocks.
They also do not reference the term "type."

dcalhoun committed:
Update Android types to match expectations

This change should've been included in an earlier commit, but was
overlooked. It updated the code utilized by the WPAndroid app to match
the rest of the Android bridging code.

dcalhoun committed:
Replace vector sparkles image with vector SVG

The monorepo configuration currently breaks serving local static images
with React Native. An inline SVG is used as a workaround for this issue
for the time being.

- https://git.io/JBV4e
- https://git.io/JBV4k

dcalhoun committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into add/native-new-block-badge

dcalhoun commented:
@mchowning @Guarani @jhnstn @kyleaparker I have addressed all feedback from the previous review. This PR and its siblings PRs are now ready for another round of review. Installable builds can be found on the WPiOS and WPAndroid PRs. 

My hope is to merge this work before the release planned for Thursday 5 July, but I understand if you have competing priorities. Thank you in advance for your time! 🙇🏻

kyleaparker commented:
Everything worked as expected for me on both Android & iOS. The only thing I noticed was some pixelation around the edge of the sparkles icon on dark mode only - do we have anyway to smooth it out? I can always redraw it as a vector icon in Figma if needed.

dcalhoun commented:
> The only thing I noticed was some pixelation around the edge of the sparkles icon on dark mode only - do we have anyway to smooth it out? I can always redraw it as a vector icon in Figma if needed.

@kyleaparker I see what you mean. Yes, I believe an improved, possibly simplified, vector asset would likely be the best way to avoid the pixelation. Would you be willing to create one?

kyleaparker commented:
> @kyleaparker I see what you mean. Yes, I believe an improved, possibly simplified, vector asset would likely be the best way to avoid the pixelation. Would you be willing to create one?

@dcalhoun yes no problem!

guarani commented on packages/editor/src/components/provider/index.native.js:
```diff
@@ -161,6 +164,24 @@ class NativeEditorProvider extends Component {
+
+		// Request current block impressions from native app
+		requestBlockTypeImpressions( ( storedImpressions ) => {
+			const impressions = { ...NEW_BLOCK_TYPES, ...storedImpressions };
```
I wanted to highlight the following scenarios with regard to updating the list of "new" blocks. I think the current behavior here is perfectly reasonable though.
- after shipping a release to users, if the need arose, we cannot remove the "sparkle" badge from a block in a subsequent version unless we make changes in the parent apps (or de-register the block completely)
- also, after shipping a release, we cannot force the "sparkle" badge to continue being shown to users on a particular block by increasing or resetting the block's impression count

dcalhoun replied:
Thank you for stating these limitations, @guarani. From my understanding of the project requirements, I do believe these are OK.

guarani reviewed (approved):
Tested on both iOS and Android and everything works as expected. I also tested a few update scenarios where I update the list of "new blocks". 
Nice work here wrangling all the platforms, @dcalhoun!

mchowning commented on packages/block-editor/src/components/inserter/hooks/use-block-type-impressions.native.js:
```diff
@@ -0,0 +1,55 @@
+			} );
+
+			// Persist block type impression count to native app store
+			setBlockTypeImpressions( updatedBlockTypeImpressions );
```
There are a couple of questions I had in giving this PR a last look. **I do not view these issues as blockers for merging this PR.** If they are changes we want to make, I think it would be better to make them in a separate PR.

One question was whether we should have to use both `updateSettings` and `setBlockTypeImpressions` every time we need to update one of the block type impressions values. It feels like we're sidestepping redux here instead of working with it. 

Did we give any thought to finding a way to have an action that changes the impressions settings (`updateSettings({impressions: ...})` kick off a call to `setBlockTypeImpressions` as a side effect? It feels like that approach would be more in keeping with the redux architecture and would avoid any risk of someone deciding to update the block impressions in some other way than using this hook and forgetting to also call `setBlockTypeImpressions` to update the value on the native side. In essence, I think what I'm wondering is whether we should be using a [controller](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-data/#controls-2) to update native anytime a relevant `updateSettings` action is dispatched.

That made me wonder if we should be doing something similar when we "select" the impressions value from the store: i.e., should the selector, when the store does not have any impressions, trigger [a resolver](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-data/#resolvers) that retrieves the impressions from native (and updates the store). In essence, it again kind of feels like by having our components firing off side effects, we're bypassing redux instead of working with it.

If we made those changes, our components would then only interact with the redux store (via actions and selectors) and all of the side effects would be handled within redux by the controllers and resolvers instead of having the side effects handled imperatively alongside the redux selectors and actions in our components. 

I'm a bit out of my depth here (I've barely touched controllers or resolvers before), so I wouldn't be at all surprised if I'm misunderstanding some things. 🙂

dcalhoun replied:
> One question was whether we should have to use both `updateSettings` and `setBlockTypeImpressions` every time we need to update one of the block type impressions values. 

I agree that managing the JavaScript and Native stores separately anytime a mutation is needed is not ideal. The current implementation was based upon prior implementations, e.g. [retrieving the editor capabilities](https://github.com/WordPress/gutenberg/blob/74a99b2cfce30a49d4c007960f24314ba1a1983e/packages/editor/src/components/provider/index.native.js#L147-L151), and discussions with the team on internal posts. I will share the internal posts separately. 

> In essence, I think what I'm wondering is whether we should be using a [controller](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-data/#controls-2) to update native anytime a relevant `updateSettings` action is dispatched.

Ah, interesting. I was not aware of `controls`. From reviewing the documentation briefly, it does seem appropriate for managing the asynchronous data flow to the Native store. 

> That made me wonder if we should be doing something similar when we "select" the impressions value from the store: i.e., should the selector, when the store does not have any impressions, trigger [a resolver](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-data/#resolvers) that retrieves the impressions from native (and updates the store).

I was also not aware of `resolvers` either. Given the description of "If your selector result may need to be fulfilled from an external source," it does seem to be a very appropriate way to manage retrieving the data from the Native store. 

> If we made those changes, our components would then only interact with the redux store (via actions and selectors) and all of the side effects would be handled within redux by the controllers and resolvers instead of having the side effects handled imperatively alongside the redux selectors and actions in our components.

Definitely. Having a single interface (i.e. the Redux store selectors and actions) to select and mutate data would be a much better approach. The `controls` and `resolvers` options do look like promising ways of accomplishing that. Thank you for sharing this information. I think it'd be great to refactor this, and possibly other implementations like the Featured Image, to leverage these options in the future.

mchowning reviewed (approved):
Left [one comment](https://github.com/WordPress/gutenberg/pull/33119#discussion_r681353716) with a few questions I had, but that comment doesn't require any changes to this PR.

This is some really great work. Thanks David! 🥇

dcalhoun committed:
Update sparkles icon with simplified vector asset

Addresses pixelation that particularly noticeable in dark mode.

jhnstn reviewed (approved):
Nice work David. I don't see anything that hasn't already been pointed out. :shipit:
